### PR TITLE
Discussion 104

### PR DESCRIPTION
**Pull request by rdeago**
This PR aims to implement all bullet points listed in #104.

I started with a little housekeeping:

- reorganized projects so all source generator projects can share code;
- created a non-packable `CommonCode` project containing all the shared code;
- made shared source files invisible in source generator projects, so they don't clutter Visual Studio's Solution Explorer;
- updated dependencies and synced files with `devlooped/oss`.

Nest steps will be:

- remove the C# 9.0 restriction - instead, check for the Roslyn version in use to determine whether ThisAssembly can be used in a project;
- make one of the source generators incremental, meanwhile working on shared code that will facilitate work on remaining generators;
- make all the other generators incremental;
- write code to generate the "main" part of the ThisAssembly class - with XML docs, attributes, access modifiers, and an empty body - and run it in _just one_ generator (the first whose .targets file is included in the project).

I will update this post as I add more commits. In the meanwhile, as of the first 13 commits ([de4993c](https://github.com/devlooped/ThisAssembly/tree/de4993ce1263cd1930edcd8a3a80c2ad2d03ab93)) the solution is buildable (at least locally, we'll see how it goes on CI) and passes all unit tests, so @kzu you may want to take a look and give some feedback on the direction I'm taking.

### CODE REVIEW

File: src/SourceGenerators/Directory.Build.props
Comment (kzu) [inline src/SourceGenerators/Directory.Build.props:8]: I'm not a fan of generalizing things like this. Doesn't save much typing and makes things harder to follow for newcomers (who won't find the TF and get confused). I'd remove this and restore the property on each project

File: src/SourceGenerators/CommonCode/CommonCode.csproj
Comment (kzu) [inline src/SourceGenerators/CommonCode/CommonCode.csproj:4]: I'd rename the folder to `Generators` (the entire project is about *source* generators anyway, and I like shorter names). Also, instead of `CommonCode`, I'd use `Common` or `Core` instead.

File: src/SourceGenerators/CommonCode/CommonCode.csproj
Comment (kzu) [inline src/SourceGenerators/CommonCode/CommonCode.csproj:14]: Yuck! Another good reason not to do away with the `TargetFramework` property, I had issue with that in the past too...

File: src/SourceGenerators/Directory.Build.targets
Comment (kzu) [inline src/SourceGenerators/Directory.Build.targets:13]: Why not just use linked files? Or directly a project reference? Not sure this is worth the added complexity... (always thinking about newcomers that might want to use this project as an example of authoring source generators or to contribute further)...

File: src/SourceGenerators/Directory.Build.targets
Comment (kzu) [inline src/SourceGenerators/Directory.Build.targets:37]: The prerequisites concept itself is somewhat superseeded by the new "common" approach here. Not sure it's worth keeping around. Say the prereq targets are directly linked and imported from each generator, we wouldn't need the dependency at all anymore... 🤔

File: src/SourceGenerators/ThisAssembly.Constants/readme.md
Comment (kzu) [inline src/SourceGenerators/ThisAssembly.Constants/readme.md:1]: Should perhaps change to `C# 9.0 or later` now that we have 10 and soon 11 :)

**Review by kzu — CHANGES_REQUESTED** [review]
Some suggestions, but it's looking great. I like the move to central package versions too. 

Thanks!

**Comment by kzu** [comment]
Hi @rdeago! Thanks a lot for taking the time to work on this. Looks very promising! 

I merged a pending PR that did the dotnet-sync update for devlooped/oss, so you can skip that commit on your PR now so it remains clean-ish.

**Comment by rdeago** [comment]
> I merged a pending PR that did the dotnet-sync update for devlooped/oss, so you can skip that commit on your PR now so it remains clean-ish.

I had already done the same thing, but no problem, I rebased the PR.

File: src/SourceGenerators/Directory.Build.props
Comment (rdeago) [reply src/SourceGenerators/Directory.Build.props]: More than the TargetFramework property itself, I was generalizing the _requisite_ that source generators target .NET Standard 2.0. I'll admit I should have made this clearer by adding a comment in Directory.Build.props pointing to relevant documentation.

Anyway, your point about newcomers still stands. Being a fan of the principle of least astonishment myself, I moved TargetFramework back into each project as suggested.

File: src/SourceGenerators/CommonCode/CommonCode.csproj
Comment (rdeago) [reply src/SourceGenerators/CommonCode/CommonCode.csproj]: Done. I chose `Common` over `Core` because the latter is widely used as a suffix for _dependencies_ containing common types, while here common types are included at a source level.

I have also moved the Common project out of the Generators directory, which simplifies Directory.Build.props as now all projects in that directory are packable. Besides, now Common can use default items, so VS stops adding a single Compile item to the project every time I add a file.

File: src/SourceGenerators/CommonCode/CommonCode.csproj
Comment (rdeago) [reply src/SourceGenerators/CommonCode/CommonCode.csproj]: I suspect the problem stems from this code in `Directory.Build.targets`:

```xml
  <PropertyGroup Condition="'$(IsPackable)' == ''">
    <!-- The Sdks\NuGet.Build.Tasks.Pack\build\NuGet.Build.Tasks.Pack.targets unconditionally sets 
        PackageId=AssemblyName if no PackageId is provided, and then defaults IsPackable=true if 
        a PackageId is set (?!), meaning that by default everything is packable in Sdk-style 
        projects. 

        The Directory.Build.targets are imported after the user's project properties have been 
        read, and therefore gives us a chance to inspect if an explicit PackageId was provided, 
        before the NuGet SDK target is imported and defaults it. At this point, we can give 
        IsPackable a more sensible default, making it false if no PackageId was provided at this 
        point. -->
    <IsPackable Condition="'$(PackageId)' == ''">false</IsPackable>
    <IsPackable Condition="'$(PackageId)' != ''">true</IsPackable>
  </PropertyGroup>
```

Maybe something has changed in the SDK so that now `IsPackable` is checked (and defaulted to `true`) before `Directory.Build.targets` is included.

I won't modify `Directory.Build.targets` because it is synced with `devlooped/oss`. Further investigation (which unfortunately I have no time to carry out at the moment) may reveal an issue in `devlooped/oss`. For now, setting `IsPackable` in the project file looks like a viable workaround.

File: src/SourceGenerators/Directory.Build.targets
Comment (rdeago) [reply src/SourceGenerators/Directory.Build.targets]: As a newcomer to the project myself, I found it annoying to have common linked files interspersed with files specific to the project. Just my opinion, of course.

To mitigate the surprise factor as much as possible, I simplified and more extensively commented the `Directory.Build.targets` file.

Given the relative complexity of the solution, an onboarding document (`contributing.md` for example) may be a good place to explain this and other peculiarities.

File: src/SourceGenerators/Directory.Build.targets
Comment (rdeago) [reply src/SourceGenerators/Directory.Build.targets]: Prerequisites as a separate package is a good place for common MSBuild code (which is bound to grow in the next few commits). It would be easy to have a `Common.targets` linked in every source generator project... except it would break the test project, as the linked file would not actually be near each project's targets file at build time. Otherwise, I agree removing the Prerequisites package would be a good thing. I'll try harder and see what I can come up with.

File: src/SourceGenerators/ThisAssembly.Constants/readme.md
Comment (rdeago) [reply src/SourceGenerators/ThisAssembly.Constants/readme.md]: The C# version used in a project is totally unrelated to the ability to use source generators. It's the version of Roslyn that counts: the latest Roslyn will happily use incremental generators in a C# 1.0 project, or even in a VB project.

I'm going to update descriptions and readmes very soon.

**Comment by kzu** [comment]
I just merged all the pending bumps. I'd rather keep this PR clean from those to avoid unrelated noise

File: src/SourceGenerators/CommonCode/CommonCode.csproj
Comment (kzu) [reply src/SourceGenerators/CommonCode/CommonCode.csproj]: Trying this in a clean project and I can't repro. So I'm guessing it might be something else...

**Comment by kzu** [comment]
Hey @rdeago this has become a monster of a PR. Very very hard to review and wrap my head around. I've been going through the code here and there, but I'm started to get the feeling that this should be several PRs entirely separate. Some obvious ones:

1. Move to central package references
2. Reorganize folders and move common stuff to new project
3. Make generators incremental
4. Masive refactoring of how code is actually generated

The last one in particular, makes the codegen much harder to follow than with simple scriban templates. Makes it very opaque to follow (for example, I was trying to figure out how the generated Constants look, and it's quite tricky to follow). I may need more time to digest the approach to determine if it's something that will be easy to maintain in the future, or if I'd rather stick to some duplication but a more straightforward approach with basic templating. I'm also not sure if the rework is a requirement of the incremental generator approach or not, I'd need to look into it some more.

Thanks again for taking the time to provide a path forward here, much appreciated!

**Comment by kzu** [comment]
Also, I'd use https://www.nuget.org/packages/isexternalinit for that particular polyfill.

**Comment by rdeago** [comment]
> Hey @rdeago this has become a monster of a PR. Very very hard to review and wrap my head around. 

Yeah, I'm sorry, my fault. I didn't anticipate it becoming such a behemoth.

> I'm started to get the feeling that this should be several PRs entirely separate. Some obvious ones:
> 
> 1. Move to central package references
> 2. Reorganize folders and move common stuff to new project
> 3. Make generators incremental
> 4. Masive refactoring of how code is actually generated

Agreed. This draft PR might remain for reference and be deleted when other PRs are merged or at least on the right track.

Central package management and folder reorganization are probably the first things to do, so Dependabot PRs and other PRs agree on which files to update. Rebasing after merging of Dependabot PRs has been a pain in the sacral region.

> [Removing embedded Scriban] makes the codegen much harder to follow than with simple scriban templates.

> I was trying to figure out how the generated Constants look, and it's quite tricky to follow

I admittedly followed a radically different approach. With templates, you look at a template and you can mostly figure out the outcome; with code generation classes, the first time you have to dig through code to find out how a Constant is translated to code, then a Class, then e.g. the Constants class. Of course when you then look at, say, AssemblyInfo, the previous cognitive burden pays up: you look at the AssemblyInfo class and you already know what the building blocks are.

I did this because it has made a lot easier to add orthogonal features, such as the possibility of generating consts vs. readonly fields. The alternative was to implement such features at the template level, essentially obtaining the same level of complexity, just duplicated four times, with more dollar signs, without syntax highlighting, and without Intellisense.

Other considerations that made me ditch Scriban include:

- Performance. You never know how many source generators a project uses, so the faster the better, within reasonable limits of course. Scriban is fast and memory-friendly [when compared to other templating engines](https://github.com/scriban/scriban/blob/master/doc/benchmarks.md), but I think it can hardly beat _not_ having templates at all.
- Since Scriban source files have no `<auto-generated/>` marker comments, they are subject to code analysis in every project that embeds them. This results in thousands (literally) of useless diagnostics, unless code analysis is turned off.

Of course, the relative weight of each of the points above is subjective, so your conclusions may wildly differ from mine.

> I may need more time to digest the approach to determine if it's something that will be easy to maintain in the future, or if I'd rather stick to some duplication but a more straightforward approach with basic templating.

By all means, take your time to assess the implications of both approaches. If you then decide in favor of templates, no hard feelings on my part: it's been an instructing and inspiring experience anyway.

> I'm also not sure if the rework is a requirement of the incremental generator approach or not

AFAIK there nothing prevents you, in principle, from using templates in an incremental generator.

> Thanks again for taking the time to provide a path forward here, much appreciated!

You're welcome! Any way you decide to go forward, I'm a @ mention away if you need me. 😉 Plus, of course, I'm still watching this PR and related discussions.

**Comment by rdeago** [comment]
> Also, I'd use https://www.nuget.org/packages/isexternalinit for that particular polyfill.

Yeah, and https://www.nuget.org/packages/indexrange for System.Index and System.Range probably. To be honest, I got fed up of looking for polyfill packages some weeks ago. Also (shameless plug ahead) I'm in the course of writing my own - with a different approach (this shouldn't surprise you by now 😁) to prevent code duplication. Long story. If you want I'll tag you in the repo as soon as it goes public.

**Comment by kzu** [comment]
> Also (shameless plug ahead) I'm in the course of writing my own - with a different approach (this shouldn't surprise you by now 😁) to prevent code duplication. Long story. If you want I'll tag you in the repo as soon as it goes public.

Please do!

**Comment by kzu** [comment]
> Central package management and folder reorganization are probably the first things to do, so Dependabot PRs and other PRs agree on which files to update. Rebasing after merging of Dependabot PRs has been a pain in the sacral region.

Agreed. That's an easy yes for merging right-away for sure.

> I did this because it makes adding orthogonal features, such as the possibility of generating consts vs. readonly fields, a lot easier. The alternative would be to implement such features at the template level, essentially obtaining the same level of complexity, just duplicated four times, with more dollar signs, without syntax highlighting, and without Intellisense.

There is another alternative: switch the template in use entirely at the generator level when you detect the setting. This will keep the template clean and simple, albeit somewhat duplicated here and there.

> * Performance. You never know how many source generators a project uses, so the faster the better, within reasonable limits of course. Scriban is fast and memory-friendly [when compared to other templating engines](https://github.com/scriban/scriban/blob/master/doc/benchmarks.md), but I think it can hardly beat _not_ having templates at all.

In my experience doing codegen for many years, I appreciate maintainability over almost everything else. So, unless there is an actual performance issue that becomes severe, I'd still value readability and maintainability over everything else. Codegen is so "meta" that unless you can (more or less) easily mentally map from model > generated code, when time comes to update the generated code (and it always does), you will have a very hard time tweaking even minor things. It's a nightmare to maintain a few years down the road when you haven't look at it for a long while (or you're a new maintainer).

> * Since Scriban source files have no `<auto-generated/>` marker comments, they are subject to code analysis in every project that embeds them. This results in thousands (literally) of useless diagnostics, unless code analysis is turned off.

I'm pretty sure this is trivial to fix. A simple .editorconfig setting can turn off code analysis for all of the templates. We could even place them all in the same folder and [turn off analysis for the entire folder](https://til.cazzulino.com/dotnet/disable-diagnostic-analyzers-for-entire-folder-submodules).

> By all means, take your time to assess the implications of both approaches. If you then decide in favor of templates, no hard feelings on my part: it's been an instructing and inspiring experience anyway.

Cool, thanks!

**Comment by kzu** [comment]
Closing this since most (all?) of it is now implemented on main. 

Thanks @rdeago for the ideas and suggestions!

**Pull request closed without merging**
